examples/platforms/nrf52840/spi-slave.c

```diff
@@ -0,0 +1,233 @@
+    NVIC_EnableIRQ(SPIS_IRQN);
+
+    /* Enable SPI slave device. */
+    nrf_spis_enable(SPIS_INSTANCE);
+
+    /* Set proper callback and context. */
+    sProcessCallback  = aProcessCallback;
+    sCompleteCallback = aCompleteCallback;
+    sContext          = aContext;
+
+exit:
+    return error;
+}
+
+void otPlatSpiSlaveDisable(void)
+{
+    /* Disable interrupts for ACQUIRED and END events. */
+    nrf_spis_int_disable(SPIS_INSTANCE, (NRF_SPIS_INT_ACQUIRED_MASK | NRF_SPIS_INT_END_MASK));
+
+    /* Disable NVIC interrupt. */
+    NVIC_DisableIRQ(SPIS_IRQN);
+
+    /* Disable SPIS instance. */
+    nrf_spis_disable(SPIS_INSTANCE);
+}
+
+otError otPlatSpiSlavePrepareTransaction(uint8_t *aOutputBuf,
+                                         uint16_t aOutputBufLen,
+                                         uint8_t *aInputBuf,
+                                         uint16_t aInputBufLen,
+                                         bool     aRequestTransactionFlag)
+{
+    otError error = OT_ERROR_NONE;
+
+    otEXPECT_ACTION(sCompleteCallback != NULL, error = OT_ERROR_INVALID_STATE);
+
+    nrf_spis_semstat_t semaphore_status = nrf_spis_semaphore_status_get(SPIS_INSTANCE);
+    otEXPECT_ACTION(((semaphore_status != NRF_SPIS_SEMSTAT_SPIS) && (semaphore_status != NRF_SPIS_SEMSTAT_CPUPENDING)),
+                    error = OT_ERROR_BUSY);
+
+    if (aOutputBuf != NULL)
+    {
+        sOutputBuf    = aOutputBuf;
+        sOutputBufLen = aOutputBufLen;
+    }
+
+    if (aInputBuf != NULL)
+    {
+        sInputBuf    = aInputBuf;
+        sInputBufLen = aInputBufLen;
+    }
+
+    sRequestTransactionFlag = aRequestTransactionFlag;
+
+    /* Trigger acquired event. */
+    nrf_spis_task_trigger(SPIS_INSTANCE, NRF_SPIS_TASK_ACQUIRE);
+
+exit:
+    return error;
+}
+
+/**
+ * Interrupt handler of SPIS peripherial.
+ */
+void SPIM0_SPIS0_TWIM0_TWIS0_SPI0_TWI0_IRQHandler(void)
+{
+    /* Check for SPI semaphore acquired event. */
+    if (nrf_spis_event_check(SPIS_INSTANCE, NRF_SPIS_EVENT_ACQUIRED))
+    {
+        nrf_spis_event_clear(SPIS_INSTANCE, NRF_SPIS_EVENT_ACQUIRED);
+
+        /* Set actual TX/RX buffers. */
+        nrf_spis_tx_buffer_set(SPIS_INSTANCE, sOutputBuf, sOutputBufLen);
+        nrf_spis_rx_buffer_set(SPIS_INSTANCE, sInputBuf, sInputBufLen);
+
+        if (sRequestTransactionFlag)
+        {
+            /* Interrupt pin is active low. */
+            nrf_gpio_pin_clear(SPIS_PIN_HOST_IRQ);
+        }
+
+        /* Trigger RELEASE event immediately. */
+        nrf_spis_task_trigger(SPIS_INSTANCE, NRF_SPIS_TASK_RELEASE);
+    }
+
+    /* Check for SPI transaction complete event. */
+    if (nrf_spis_event_check(SPIS_INSTANCE, NRF_SPIS_EVENT_END))
+    {
+        nrf_spis_event_clear(SPIS_INSTANCE, NRF_SPIS_EVENT_END);
+
+        if (sRequestTransactionFlag)
+        {
+            nrf_gpio_pin_set(SPIS_PIN_HOST_IRQ);
+        }
+
+        /* Set transaction done flag. */
+        sTransactionDone = true;
```

Comment:
The concept behind the spi-slave platform  transaction `complete` and `process` callbacks is like this:
- Preferably the transaction done callback `sCompleteCallback` is called from ISR so that the `NcpSpi` can immediately prepare any next transaction  (set the input output buffer)... This is to ensure that we have a properly formatted response if master starts a new transaction.
- If there are further actions required `sCompleteCallback` would return `true` and then platform would need to call `sProcessCallback` (from the same OS context that OpenThread is running) to allow OT and `NcpSpi` to do any further processing.

Two questions/suggestions: 
- Can we consider invoking `sCompleteCallback` from ISR here? The `NcpSpi` would then prepare any next transaction by call into `otPlatSpiSlavePrepareTransaction()`....
- Would `otPlatSpiSlavePrepareTransaction()` implementation allow it to be called safely called from the interrupt context (and would it then update the buffers )?

Reply:
You are right, I will modify the current implementation.

Reply:
Thanks